src/ContentCreator.php

```diff
@@ -0,0 +1,858 @@
+   *
+   * @var array
+   */
+  protected $globalState = [];
+
+  /**
+   * Keeps stack of entity referencing each other, to avoid infinite loop.
+   *
+   * @var array
+   */
+  protected $entityTypeReferenceNestingStack = [];
+
+  /**
+   * List of entity types to create.
+   *
+   * @var array
+   */
+  protected $baseEntityTypes = [
+    'node',
+  ];
+
+  /**
+   * The content creator storage service.
+   *
+   * @var \Drupal\testsite_builder\ContentCreatorStorage
+   */
+  protected $storage;
+
+  /**
+   * Constructs a new ContentCreator object.
+   *
+   * @param \Drupal\testsite_builder\ContentCreatorStorage $storage
+   *   The content creator storage service.
+   * @param \Drupal\Core\Database\Connection $database
+   *   The database connection.
+   * @param \Drupal\testsite_builder\BaseEntityTablesPluginManager $base_entity_tables_plugin_manager
+   *   The base entity tables plugin manager service.
+   */
+  public function __construct(ContentCreatorStorage $storage, Connection $database, BaseEntityTablesPluginManager $base_entity_tables_plugin_manager) {
+    $this->storage = $storage;
+    $this->database = $database;
+    $this->baseEntityTablesPluginManager = $base_entity_tables_plugin_manager;
+  }
+
+  /**
+   * Init global state.
+   */
+  protected function init() {
+    $this->config = $this->storage->getConfig();
+    $this->sampleDataTypes = $this->storage->getSampleData();
+
+    $this->outputDirectory = sys_get_temp_dir() . '/' . uniqid('testsite_builder_content_creator_', TRUE);
+    mkdir($this->outputDirectory, 0777, TRUE);
+
+    // Store config and sample data.
+    file_put_contents($this->outputDirectory . '/_config.json', json_encode($this->config, JSON_PRETTY_PRINT));
+    file_put_contents($this->outputDirectory . '/_sample_data.json', json_encode($this->sampleDataTypes, JSON_PRETTY_PRINT));
+
+    $this->globalState = [
+      'entity_index' => 0,
+      'bundle_index' => 0,
+      'entity_states' => [],
+    ];
+  }
+
+  /**
+   * Init caching values for entity type.
+   *
+   * @param string $entity_type
+   *   The entity type.
+   */
+  protected function initEntityState($entity_type) {
+    if (isset($this->globalState['entity_states'][$entity_type])) {
+      return;
+    }
+
+    $this->globalState['entity_states'][$entity_type]['count'] = 1;
+    $this->globalState['entity_states'][$entity_type]['index'] = $this->globalState['entity_index'];
+    $this->globalState['entity_index']++;
+  }
+
+  /**
+   * Init caching values for bundle type.
+   *
+   * @param string $entity_type
+   *   The entity type.
+   * @param string $bundle_type
+   *   The bundle type.
+   */
+  protected function initBundleState($entity_type, $bundle_type) {
+    if (isset($this->globalState['entity_states'][$entity_type][$bundle_type])) {
+      return;
+    }
+
+    $this->globalState['entity_states'][$entity_type][$bundle_type]['index'] = $this->globalState['bundle_index'];
+    $this->globalState['bundle_index']++;
```

Comment:
The bundle index afaics doesn't need to be for all entity types so you could do `count($this->globalState['entity_states'][$entity_type][$bundle_type])` as well.

Also I wouldn't call these keys `index` because it is a bit confusing - how about `uuid_part` or something similar.

Reply:
Also once we've removed both `entity_index` and `bundle_index` we can change the name of $globalState to $entityState or maybe $entityCounts and move everything up one.